src/attributes/testing.md

```diff
@@ -9,26 +9,25 @@ enables the [`test` conditional compilation option].
 
 The *`test` attribute* marks a function to be executed as a test. These
 functions are only compiled when in test mode. Test functions must be free,
-monomorphic functions that take no arguments, and the return type must be one
-of the following:
+monomorphic functions that take no arguments, and the return type must implement the [`Termination`] trait, for example:
 
 * `()`
-* `Result<(), E> where E: Error`
-<!-- * `!` -->
-<!-- * Result<!, E> where E: Error` -->
-
-> Note: The implementation of which return types are allowed is determined by
-> the unstable [`Termination`] trait.
+* `Result<(), E> where E: Debug`
+* `!`
+<!-- * Result<!, E> where E: Debug` -->
 
 <!-- If the previous section needs updating (from "must take no arguments"
   onwards, also update it in the crates-and-source-files.md file -->
 
 > Note: The test mode is enabled by passing the `--test` argument to `rustc`
 > or using `cargo test`.
 
-Tests that return `()` pass as long as they terminate and do not panic. Tests
-that return a `Result<(), E>` pass as long as they return `Ok(())`. Tests that
-do not terminate neither pass nor fail.
+The test harness calls the returned value's [`report`] method, and classifies the test as passed or failed depending on whether the resulting [`ExitCode`] represents successful termination.
+In particular:
+* Tests that return `()` pass as long as they terminate and do not panic.
+* Tests that return a `Result<(), E>` pass as long as they return `Ok(())`.
+* Tests that return `ExitCode::SUCCESS` pass, and tests that return `ExitCode::FAILURE` fail.
+* Tests that do not terminate neither pass nor fail.
```

Comment:
I'm concerned this may be over-specifying the behavior. There's a bit of an awkward separation between "Rust the language" and "the standard library".  The reference typically tends to focus on just the language behavior, and not delve into the standard library's implementation.

Also, I'm not really comfortable with specifying the non-termination behavior. I would consider that a failure, and it is possible in the future that either libtest or cargo could have some sort of timeout controls which would consider the test to fail.

I think leaving this up to the `Termination` docs in the standard library should be sufficient.

Reply:
The "Tests that do not terminate neither pass nor fail." sentence was present already; this PR only moves it.